src/poetry/core/packages/constraints/any_constraint.py

```diff
@@ -1,33 +1,27 @@
-from typing import TYPE_CHECKING
-
 from poetry.core.packages.constraints.base_constraint import BaseConstraint
 from poetry.core.packages.constraints.empty_constraint import EmptyConstraint
 
 
-if TYPE_CHECKING:
-    from poetry.core.packages.constraints import ConstraintTypes  # noqa
-
-
 class AnyConstraint(BaseConstraint):
-    def allows(self, other: "ConstraintTypes") -> bool:
+    def allows(self, other: "BaseConstraint") -> bool:
         return True
 
-    def allows_all(self, other: "ConstraintTypes") -> bool:
+    def allows_all(self, other: "BaseConstraint") -> bool:
         return True
 
-    def allows_any(self, other: "ConstraintTypes") -> bool:
+    def allows_any(self, other: "BaseConstraint") -> bool:
         return True
 
-    def difference(self, other: "ConstraintTypes") -> "ConstraintTypes":
+    def difference(self, other: "BaseConstraint") -> "BaseConstraint":
         if other.is_any():
             return EmptyConstraint()
 
         return other
 
-    def intersect(self, other: "ConstraintTypes") -> "ConstraintTypes":
+    def intersect(self, other: "BaseConstraint") -> "BaseConstraint":
```

Comment:
Hey,

just a quick comment on that. While returning "BaseConstraint" is technical correct, this might lead missing code completion and "Unresolved attribute errors" in the IDE, because the several Constraint Types have different additional attributes. That's the reason why we used "ConstraintTypes" until now. Another solution to this, is to typecast the return values of these methods in the place where they were called.

fin swimmer

Reply:
Seeing as how you don't know which type of constraint you have in hand, you mustn't rely on any of those completions until you have refined (by casting or isinstance or whatever) - right?

ie it only leads to missing completions in places where completions _should_ be missing